### PR TITLE
Add a configurable webhook execution delay

### PR DESCRIPTION
**Pull request by hectorgomezv**
Changes:
- Adds a configurable (via the environment variable `WEB_HOOK_EXECUTION_DELAY_MS`) delay before executing web hook. The default value for this is 5 seconds.
- The status code returned by the controller was changed to 202 as it is more accurate given the new behavior.

### CODE REVIEW

File: src/routes/cache-hooks/cache-hooks.controller.ts [hunk 1]
```diff
@@ -43,6 +43,6 @@ export class CacheHooksController {
       | PendingTransaction
       | SafeAppsUpdate,
   ): Promise<void> {
-    await this.service.onEvent(eventPayload);
+    this.service.onEvent(eventPayload);
```
Comment (hectorgomezv) [inline src/routes/cache-hooks/cache-hooks.controller.ts:46]: This is changed to run the `onEvent` callback in a non-blocking fashion and avoid having the caller waiting for the event hook execution.

Comment (fmrsabino) [reply src/routes/cache-hooks/cache-hooks.controller.ts:46]: I think it'd be simpler if the delay was tackled on a web-hook route only initially. So you'd have something like

```typescript
setTimeout(() => {
  this.service.onEvent(eventPayload);
}, 5000); // read timeout value from the configuration
```

I would also add a small test to this where the `service` is mocked and we use virtual timers. With the mock you can check that the service was only called after T seconds/milliseconds. Due to the scope of this issue it's also fine if the test is added as a separate PR 👍

Comment (hectorgomezv) [reply src/routes/cache-hooks/cache-hooks.controller.ts]: I've changed it in [563547d](https://github.com/safe-global/safe-client-gateway/pull/883/commits/563547d36d6c47eb6cdbbab2242173d5723f4fbf)

I had to adjust the controller tests in this PR, they would fail otherwise, as `fakeCacheService.get` wouldn't return `null` immediately as the controller tests were expecting.

Comment (hectorgomezv) [reply src/routes/cache-hooks/cache-hooks.controller.ts]: I also changed the status code returned by the controller to `202` as it is more accurate given the new behavior.

**Comment by hectorgomezv** [comment]
The status code returned by the controller was changed to `202` as it is more accurate given the new behavior. 

The PR title and description were also changed.

File: src/routes/cache-hooks/cache-hooks.controller.ts [hunk 2]
```diff
@@ -15,18 +22,29 @@ import { PendingTransaction } from '@/routes/cache-hooks/entities/pending-transa
...
+  ) {
+    this.webHookExecutionDelayMs = this.configurationService.getOrThrow<number>(
+      'webHookExecutionDelayMs',
+    );
+  }
 
   @UseGuards(BasicAuthGuard)
   @Post('/hooks/events')
-  @HttpCode(200)
+  @HttpCode(202)
```
Comment (fmrsabino) [inline src/routes/cache-hooks/cache-hooks.controller.ts:47]: What's the reason for this change? Seems to be out of context or? 🤔

File: src/routes/cache-hooks/cache-hooks.controller.spec.ts [hunk 3]
```diff
@@ -407,9 +428,12 @@ describe('Post Hook Events (Unit)', () => {
       .post(`/hooks/events`)
       .set('Authorization', `Basic ${authToken}`)
       .send(data)
-      .expect(200);
+      .expect(202);
 
-    await expect(fakeCacheService.get(cacheDir)).resolves.toBeNull();
+    setTimeout(
+      () => expect(fakeCacheService.get(cacheDir)).resolves.toBeNull(),
+      webHookExecutionDelayMs,
```
Comment (fmrsabino) [inline src/routes/cache-hooks/cache-hooks.controller.spec.ts:435]: This would add the assertion to the event loop which I don't think it's guaranteed to work after `webHookExecutionDelayMs`.

Did you try to use with `jest.advanceTimersByTime`?

Comment (hectorgomezv) [reply src/routes/cache-hooks/cache-hooks.controller.spec.ts:435]: > This would add the assertion to the event loop which I don't think it's guaranteed to work after `webHookExecutionDelayMs`.

Could you elaborate on why this is not guaranteed to work? 

> Did you try to use with `jest.advanceTimersByTime`?

I saw it in the documentation, but it only works in the context of jest mocks, right? `fakeCacheService` is not a jest mock in this class. I'm not following on how to use that here, could you elaborate on that?

Comment (fmrsabino) [reply src/routes/cache-hooks/cache-hooks.controller.spec.ts:435]: > Could you elaborate on why this is not guaranteed to work?

The assertion will be added to the event loop but that doesn't mean that it will be executed exactly at `webHookExecutionDelayMs`. It will be executed at the next event cycle. I believe there can be a use case where the deletion can happen after the assertion. 
 
> I saw it in the documentation, but it only works in the context of jest mocks, right? `fakeCacheService` is not a jest mock in this class. I'm not following on how to use that here, could you elaborate on that?

Not really. Jest replaces the timers with fake ones. It is not related with mocks. See https://jestjs.io/docs/timer-mocks

**Review by fmrsabino — APPROVED** [review]
We should revisit the way the timers work in these tests or have one where the the value retrieved from cache is still present and not null.

🚀

File: src/routes/cache-hooks/cache-hooks.controller.spec.ts (hunk 3 above)
Comment (hectorgomezv) [reply src/routes/cache-hooks/cache-hooks.controller.spec.ts:435]: Synced internally, thanks for your comments @fmrsabino ! 🙂
